File: Packages/StrandAnalytics/Sources/StrandAnalytics/SafetyCheckInPolicy.swift
import Foundation

// SafetyCheckInPolicy.swift - pure policy and message construction for NOOP's manual Safety Center.
//
// This is deliberately NOT a fall detector, medical-event detector, dispatcher, or background delivery
// service. It only (a) describes the state of a user-armed local reminder and (b) prepares text the user
// may choose to share through the operating system. The platform layer owns permissions, notifications,
// location capture, calling, and the final user-confirmed send action.
//
// Keep this byte-for-byte behavior aligned with Android SafetyCheckInPolicy.kt.

public enum SafetyShareIntent: String, CaseIterable, Equatable, Sendable {
    case needHelpNow
    case feelUnsafe
    case missedCheckIn
}

/// Platform-supplied copy for the user-confirmed safety draft.
///
/// The pure policy keeps an English default for deterministic tests and headless callers. App targets
/// pass localized values at the final presentation boundary so a translated Safety Center does not
/// unexpectedly open an English-only message.
public struct SafetyShareCopy: Equatable, Sendable {
    public let needHelpNowOpening: String
    public let feelUnsafeOpening: String
    public let missedCheckInOpening: String
    public let immediateDangerInstruction: String
    public let locationLabel: String
    public let locationCapturedFormat: String
    public let locationCapturedAccuracyFormat: String
    public let noteLabel: String
    public let preparedAtFormat: String
    public let deliveryBoundary: String

    public init(needHelpNowOpening: String,
                feelUnsafeOpening: String,
                missedCheckInOpening: String,
                immediateDangerInstruction: String,
                locationLabel: String,
                locationCapturedFormat: String,
                locationCapturedAccuracyFormat: String,
                noteLabel: String,
                preparedAtFormat: String,
                deliveryBoundary: String) {
        self.needHelpNowOpening = needHelpNowOpening
        self.feelUnsafeOpening = feelUnsafeOpening
        self.missedCheckInOpening = missedCheckInOpening
        self.immediateDangerInstruction = immediateDangerInstruction
        self.locationLabel = locationLabel
        self.locationCapturedFormat = locationCapturedFormat
        self.locationCapturedAccuracyFormat = locationCapturedAccuracyFormat
        self.noteLabel = noteLabel
        self.preparedAtFormat = preparedAtFormat
        self.deliveryBoundary = deliveryBoundary
    }

    public static let english = SafetyShareCopy(
        needHelpNowOpening: "I need help now. Please call me.",
        feelUnsafeOpening: "I feel unsafe. Please call me and stay on the line if you can.",
        missedCheckInOpening: "I missed a planned check-in. Please contact me.",
        immediateDangerInstruction:
            "If you think I am in immediate danger, contact local emergency services.",
        locationLabel: "Location",
        locationCapturedFormat: "Location captured at %1$@.",
        locationCapturedAccuracyFormat:
            "Location captured at %1$@, accuracy about %2$lld m.",
        noteLabel: "Note",
        preparedAtFormat: "Prepared at %1$@.",
        deliveryBoundary:
            "NOOP did not send this automatically and does not monitor or contact emergency services."
    )

    fileprivate func opening(for intent: SafetyShareIntent) -> String {
        switch intent {
        case .needHelpNow: return needHelpNowOpening
        case .feelUnsafe: return feelUnsafeOpening
        case .missedCheckIn: return missedCheckInOpening
        }
    }
}

public struct SafetyLocation: Equatable, Sendable {
    /// A user-requested one-shot fix must stay recent enough to describe where they are now. Both
    /// platforms can return cached locations, so structural coordinate validation is not sufficient.
    public static let maximumAgeSeconds = 5 * 60
    public static let maximumFutureClockSkewSeconds = 60

    public let latitude: Double
    public let longitude: Double
    public let horizontalAccuracyMeters: Double?
    public let capturedAtUnix: Int

    public init(latitude: Double,
                longitude: Double,
                horizontalAccuracyMeters: Double? = nil,
                capturedAtUnix: Int) {
        self.latitude = latitude
        self.longitude = longitude
        self.horizontalAccuracyMeters = horizontalAccuracyMeters
        self.capturedAtUnix = capturedAtUnix
    }

    public var isValid: Bool {
        latitude.isFinite && longitude.isFinite
            && (-90.0...90.0).contains(latitude)
            && (-180.0...180.0).contains(longitude)
            && capturedAtUnix > 0
    }

    public func isUsable(atUnix nowUnix: Int) -> Bool {
        guard isValid, nowUnix > 0 else { return false }
        let (age, overflow) = nowUnix.subtractingReportingOverflow(capturedAtUnix)
        guard !overflow else { return false }
        return age >= -Self.maximumFutureClockSkewSeconds
            && age <= Self.maximumAgeSeconds
    }
}

public enum SafetyShareMessage {
    public static let maxNameCharacters = 80
    public static let maxNoteCharacters = 240

    /// Build a compact, user-shareable message. Invalid coordinates are omitted rather than guessed.
    /// The final sentence is load-bearing safety copy: NOOP prepared the draft but did not dispatch it.
    public static func build(intent: SafetyShareIntent,
                             displayName: String? = nil,
                             note: String? = nil,
                             location: SafetyLocation? = nil,
                             preparedAtUnix: Int,
                             copy: SafetyShareCopy = .english) -> String {
        let cleanName = clean(displayName, maxCharacters: maxNameCharacters)
        let cleanNote = clean(note, maxCharacters: maxNoteCharacters)
        var parts: [String] = []
        let opening = copy.opening(for: intent)

        if let cleanName {
            parts.append("\(cleanName): \(opening)")
        } else {
            parts.append(opening)
        }

        parts.append(copy.immediateDangerInstruction)

        if let location, location.isUsable(atUnix: preparedAtUnix) {
            parts.append("\(copy.locationLabel): \(mapURL(location))")
            let accuracy = location.horizontalAccuracyMeters
                .flatMap { $0.isFinite && $0 >= 0 ? Int($0.rounded()) : nil }
            if let accuracy {
                parts.append(String(
                    format: copy.locationCapturedAccuracyFormat,
                    locale: Locale.current,
                    arguments: [utcTimestamp(location.capturedAtUnix), Int64(accuracy)]
                ))
            } else {
                parts.append(String(
                    format: copy.locationCapturedFormat,
                    locale: Locale.current,
                    arguments: [utcTimestamp(location.capturedAtUnix)]
                ))
            }
        }

        if let cleanNote {
            parts.append("\(copy.noteLabel): \(cleanNote)")
        }

        if preparedAtUnix > 0 {
            parts.append(String(
                format: copy.preparedAtFormat,
                locale: Locale.current,
                arguments: [utcTimestamp(preparedAtUnix)]
            ))
        }
        parts.append(copy.deliveryBoundary)
        return parts.joined(separator: "\n")
    }

    static func mapURL(_ location: SafetyLocation) -> String {
        let latitude = String(format: "%.6f", locale: Locale(identifier: "en_US_POSIX"), location.latitude)
        let longitude = String(format: "%.6f", locale: Locale(identifier: "en_US_POSIX"), location.longitude)
        return "https://www.google.com/maps/search/?api=1&query=\(latitude),\(longitude)"
    }

    static func utcTimestamp(_ unix: Int) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: Date(timeIntervalSince1970: TimeInterval(unix)))
    }

    /// Remove controls/newlines, collapse whitespace, trim, and cap by Unicode characters.
    static func clean(_ raw: String?, maxCharacters: Int) -> String? {
        guard let raw else { return nil }
        let normalized = raw.unicodeScalars.map { scalar -> Character in
            Character(scalar.properties.isWhitespace || scalar.properties.generalCategory == .control ? " " : scalar)
        }
        let collapsed = String(normalized)
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
        guard !collapsed.isEmpty else { return nil }
        return String(collapsed.prefix(max(0, maxCharacters)))
    }
}

public enum SafetyCheckInState: Equatable, Sendable {
    case inactive
    case active(remainingSeconds: Int)
    case dueSoon(remainingSeconds: Int)
    case overdue(elapsedSeconds: Int)
}

public enum SafetyCheckInPolicy {
    public static let minimumDurationSeconds = 5 * 60
    public static let maximumDurationSeconds = 24 * 60 * 60
    public static let dueSoonWindowSeconds = 10 * 60

    public static let notificationTitle = "Personal check-in"
    public static let notificationBody =
        "Your timer ended. Check in with someone you trust if you still need to."

    public static func clampedDurationSeconds(_ requested: Int) -> Int {
        min(max(requested, minimumDurationSeconds), maximumDurationSeconds)
    }

    public static func dueAtUnix(startedAtUnix: Int, requestedDurationSeconds: Int) -> Int? {
        guard startedAtUnix > 0 else { return nil }
        let duration = clampedDurationSeconds(requestedDurationSeconds)
        let (due, overflow) = startedAtUnix.addingReportingOverflow(duration)
        return overflow ? nil : due
    }

    /// Resolve current state from a persisted due timestamp. The reminder is local-only; overdue does
    /// not imply that a contact or emergency service was notified.
    public static func state(dueAtUnix: Int?, nowUnix: Int) -> SafetyCheckInState {
        guard let dueAtUnix, dueAtUnix > 0, nowUnix > 0 else { return .inactive }
        let remaining = dueAtUnix - nowUnix
        if remaining <= 0 {
            return .overdue(elapsedSeconds: max(0, -remaining))
        }
        if remaining <= dueSoonWindowSeconds {
            return .dueSoon(remainingSeconds: remaining)
        }
        return .active(remainingSeconds: remaining)
    }

    public static func statusLabel(_ state: SafetyCheckInState) -> String {
        switch state {
        case .inactive:
            return "No check-in timer is active."
        case let .active(remainingSeconds):
            return "Check-in due in \(durationLabel(remainingSeconds))."
        case let .dueSoon(remainingSeconds):
            return "Check-in due soon: \(durationLabel(remainingSeconds)) remaining."
        case let .overdue(elapsedSeconds):
            return "Check-in overdue by \(durationLabel(elapsedSeconds)). No message was sent automatically."
        }
    }

    static func durationLabel(_ seconds: Int) -> String {
        let safe = max(0, seconds)
        if safe < 60 { return "\(safe)s" }
        if safe < 3_600 { return "\(safe / 60)m" }
        let hours = safe / 3_600
        let minutes = (safe % 3_600) / 60
        return minutes == 0 ? "\(hours)h" : "\(hours)h \(minutes)m"
    }
}

File: android/app/src/main/java/com/noop/safety/SafetyCheckInPolicy.kt
package com.noop.safety

import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.math.roundToInt

/**
 * Pure policy and message construction for NOOP's manual Safety Center.
 *
 * This is not a fall detector, medical-event detector, dispatcher, or background delivery service. It
 * only describes a user-armed local reminder and prepares text the user may choose to share through the
 * operating system. Keep behavior aligned with Swift SafetyCheckInPolicy.swift.
 */
enum class SafetyShareIntent {
    NEED_HELP_NOW,
    FEEL_UNSAFE,
    MISSED_CHECK_IN,
}

/** Platform-supplied copy for the user-confirmed safety draft. */
data class SafetyShareCopy(
    val needHelpNowOpening: String,
    val feelUnsafeOpening: String,
    val missedCheckInOpening: String,
    val immediateDangerInstruction: String,
    val locationLabel: String,
    val locationCapturedFormat: String,
    val locationCapturedAccuracyFormat: String,
    val noteLabel: String,
    val preparedAtFormat: String,
    val deliveryBoundary: String,
) {
    fun opening(intent: SafetyShareIntent): String = when (intent) {
        SafetyShareIntent.NEED_HELP_NOW -> needHelpNowOpening
        SafetyShareIntent.FEEL_UNSAFE -> feelUnsafeOpening
        SafetyShareIntent.MISSED_CHECK_IN -> missedCheckInOpening
    }

    companion object {
        val English = SafetyShareCopy(
            needHelpNowOpening = "I need help now. Please call me.",
            feelUnsafeOpening = "I feel unsafe. Please call me and stay on the line if you can.",
            missedCheckInOpening = "I missed a planned check-in. Please contact me.",
            immediateDangerInstruction =
                "If you think I am in immediate danger, contact local emergency services.",
            locationLabel = "Location",
            locationCapturedFormat = "Location captured at %1\$s.",
            locationCapturedAccuracyFormat =
                "Location captured at %1\$s, accuracy about %2\$d m.",
            noteLabel = "Note",
            preparedAtFormat = "Prepared at %1\$s.",
            deliveryBoundary =
                "NOOP did not send this automatically and does not monitor or contact emergency services.",
        )
    }
}

data class SafetyLocation(
    val latitude: Double,
    val longitude: Double,
    val horizontalAccuracyMeters: Double? = null,
    val capturedAtUnix: Long,
) {
    companion object {
        const val MAXIMUM_AGE_SECONDS = 5 * 60L
        const val MAXIMUM_FUTURE_CLOCK_SKEW_SECONDS = 60L
    }

    val isValid: Boolean
        get() = latitude.isFinite() && longitude.isFinite() &&
            latitude in -90.0..90.0 && longitude in -180.0..180.0 &&
            capturedAtUnix > 0L

    fun isUsable(atUnix: Long): Boolean {
        if (!isValid || atUnix <= 0L) return false
        val age = atUnix - capturedAtUnix
        return age >= -MAXIMUM_FUTURE_CLOCK_SKEW_SECONDS && age <= MAXIMUM_AGE_SECONDS
    }
}

object SafetyShareMessage {
    const val MAX_NAME_CHARACTERS = 80
    const val MAX_NOTE_CHARACTERS = 240

    fun build(
        intent: SafetyShareIntent,
        displayName: String? = null,
        note: String? = null,
        location: SafetyLocation? = null,
        preparedAtUnix: Long,
        copy: SafetyShareCopy = SafetyShareCopy.English,
    ): String {
        val cleanName = clean(displayName, MAX_NAME_CHARACTERS)
        val cleanNote = clean(note, MAX_NOTE_CHARACTERS)
        val parts = mutableListOf<String>()
        val opening = copy.opening(intent)

        parts += if (cleanName != null) "$cleanName: $opening" else opening
        parts += copy.immediateDangerInstruction

        if (location?.isUsable(preparedAtUnix) == true) {
            parts += "${copy.locationLabel}: ${mapUrl(location)}"
            val accuracy = location.horizontalAccuracyMeters
                ?.takeIf { it.isFinite() && it >= 0.0 }
                ?.roundToInt()
            parts += if (accuracy != null) {
                String.format(
                    Locale.getDefault(),
                    copy.locationCapturedAccuracyFormat,
                    utcTimestamp(location.capturedAtUnix),
                    accuracy,
                )
            } else {
                String.format(
                    Locale.getDefault(),
                    copy.locationCapturedFormat,
                    utcTimestamp(location.capturedAtUnix),
                )
            }
        }

        if (cleanNote != null) parts += "${copy.noteLabel}: $cleanNote"
        if (preparedAtUnix > 0L) {
            parts += String.format(
                Locale.getDefault(),
                copy.preparedAtFormat,
                utcTimestamp(preparedAtUnix),
            )
        }
        parts += copy.deliveryBoundary
        return parts.joinToString("\n")
    }

    internal fun mapUrl(location: SafetyLocation): String {
        val latitude = String.format(Locale.US, "%.6f", location.latitude)
        val longitude = String.format(Locale.US, "%.6f", location.longitude)
        return "https://www.google.com/maps/search/?api=1&query=$latitude,$longitude"
    }

    internal fun utcTimestamp(unix: Long): String =
        DateTimeFormatter.ISO_INSTANT.format(Instant.ofEpochSecond(unix))

    internal fun clean(raw: String?, maxCharacters: Int): String? {
        if (raw == null) return null
        val collapsed = raw
            .map { if (it.isWhitespace() || it.isISOControl()) ' ' else it }
            .joinToString("")
            .trim()
            .split(Regex("\\s+"))
            .filter { it.isNotEmpty() }
            .joinToString(" ")
        if (collapsed.isEmpty()) return null
        return collapsed.take(maxCharacters.coerceAtLeast(0))
    }
}

sealed class SafetyCheckInState {
    data object Inactive : SafetyCheckInState()
    data class Active(val remainingSeconds: Long) : SafetyCheckInState()
    data class DueSoon(val remainingSeconds: Long) : SafetyCheckInState()
    data class Overdue(val elapsedSeconds: Long) : SafetyCheckInState()
}

object SafetyCheckInPolicy {
    const val MINIMUM_DURATION_SECONDS = 5 * 60L
    const val MAXIMUM_DURATION_SECONDS = 24 * 60 * 60L
    const val DUE_SOON_WINDOW_SECONDS = 10 * 60L

    const val NOTIFICATION_TITLE = "Personal check-in"
    const val NOTIFICATION_BODY =
        "Your timer ended. Check in with someone you trust if you still need to."

    fun clampedDurationSeconds(requested: Long): Long =
        requested.coerceIn(MINIMUM_DURATION_SECONDS, MAXIMUM_DURATION_SECONDS)

    fun dueAtUnix(startedAtUnix: Long, requestedDurationSeconds: Long): Long? {
        if (startedAtUnix <= 0L) return null
        val duration = clampedDurationSeconds(requestedDurationSeconds)
        if (startedAtUnix > Long.MAX_VALUE - duration) return null
        return startedAtUnix + duration
    }

    fun state(dueAtUnix: Long?, nowUnix: Long): SafetyCheckInState {
        if (dueAtUnix == null || dueAtUnix <= 0L || nowUnix <= 0L) return SafetyCheckInState.Inactive
        val remaining = dueAtUnix - nowUnix
        return when {
            remaining <= 0L -> SafetyCheckInState.Overdue((-remaining).coerceAtLeast(0L))
            remaining <= DUE_SOON_WINDOW_SECONDS -> SafetyCheckInState.DueSoon(remaining)
            else -> SafetyCheckInState.Active(remaining)
        }
    }

    fun statusLabel(state: SafetyCheckInState): String = when (state) {
        SafetyCheckInState.Inactive -> "No check-in timer is active."
        is SafetyCheckInState.Active -> "Check-in due in ${durationLabel(state.remainingSeconds)}."
        is SafetyCheckInState.DueSoon ->
            "Check-in due soon: ${durationLabel(state.remainingSeconds)} remaining."
        is SafetyCheckInState.Overdue ->
            "Check-in overdue by ${durationLabel(state.elapsedSeconds)}. No message was sent automatically."
    }

    internal fun durationLabel(seconds: Long): String {
        val safe = seconds.coerceAtLeast(0L)
        if (safe < 60L) return "${safe}s"
        if (safe < 3_600L) return "${safe / 60L}m"
        val hours = safe / 3_600L
        val minutes = (safe % 3_600L) / 60L
        return if (minutes == 0L) "${hours}h" else "${hours}h ${minutes}m"
    }
}
